Sources/SwiftProtobuf/Message+BinaryAdditions.swift

```diff
@@ -0,0 +1,133 @@
+// Sources/SwiftProtobuf/BinaryTypeAdditions.swift - Per-type binary coding
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/master/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extensions to the proto types defined in ProtobufTypes.swift to provide
+/// type-specific binary coding and decoding.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Binary encoding and decoding methods for messages.
+public extension Message {
+  /// Returns a `Data` value containing the Protocol Buffer binary format
+  /// serialization of the message.
+  ///
+  /// - Parameters:
+  ///   - partial: Binary serialization requires all `required` fields be
+  ///     present; if `false` (the defaul), this method throws
+  ///     `BinaryEncodingError.missingRequiredFields` if any were missing. When
+  ///     `partial` is `true`, then partial messages are allowed and
+  ///     `Message.isInitialized` is not checked.
+  /// - Returns: A `Data` value containing the binary serialization of the
+  ///   message.
+  /// - Throws: `BinaryEncodingError` if encoding fails.
+  func serializedData(partial: Bool = false) throws -> Data {
+    if !partial && !isInitialized {
+      throw BinaryEncodingError.missingRequiredFields
+    }
+    let requiredSize = try serializedDataSize()
+    var data = Data(count: requiredSize)
+    try data.withUnsafeMutableBytes { (pointer: UnsafeMutablePointer<UInt8>) in
+      try serializeBinary(into: pointer)
+    }
+    return data
+  }
+
+  private func serializeBinary(
+    into pointer: UnsafeMutablePointer<UInt8>
+  ) throws {
+    var visitor = BinaryEncodingVisitor(forWritingInto: pointer)
+    try traverse(visitor: &visitor)
+  }
+
+  internal func serializedDataSize() throws -> Int {
```

Comment:
I personally would include basic documentation here for the benefit of anyone who wants to work on the code someday.  How about
```
/// (Internal API) Computes the size of the binary serialized data.
///
/// This is used by `serializedData()` to determine the correct size of the
/// output buffer so that the low-level encoding can proceed at full speed without bounds checks.
```